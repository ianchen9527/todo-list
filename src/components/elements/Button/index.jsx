import React from "react";
import styles from "./index.module.scss";
import PropTypes from "prop-types";
import cx from "classnames";

const getWrapperClassName = (disabled) =>
  cx({
    [styles["wrapper"]]: true,
    [styles["disabled"]]: disabled,
  });

const Button = ({ icon, loading, onClick, text, disabled }) => {
  return (
    <div className={getWrapperClassName(disabled)} onClick={onClick}>
      {icon && <div className={styles[`icon-${icon}`]} />}
      {text}
      {loading && <div className={styles["loading"]} />}
    </div>
  );
};

Button.propTypes = {
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  icon: PropTypes.string,
  text: PropTypes.string,
};

Button.defaultProps = {
  onClick: null,
  diabled: false,
  loading: false,
  icon: "add",
  text: "",
};

export default Button;
