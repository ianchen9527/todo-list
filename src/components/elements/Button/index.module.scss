@import "../../../stylesheets/constants/colors.scss";

.wrapper {
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  align-content: center;
  height: 40px;
  background-color: $button;
  border: 1px solid $black;
  border-radius: 5px;
  cursor: pointer;
  color: $black;
  font-family: "Roboto";
  font-size: 14px;
  line-height: 21px;

  &:hover {
    background-color: $button-hover;
  }

  &.disabled {
    opacity: 0.5;
    cursor: default;

    &:hover {
      background-color: $button;
    }
  }

  @mixin icon {
    width: 20px;
    height: 20px;
    display: block;
    background-size: 20px 20px;
    margin-right: 15px;
  }

  .icon-add {
    @include icon();
    
    background-image: url("../../../assets/images/add.svg");
  }

  .icon-edit {
    @include icon();
    
    background-image: url("../../../assets/images/edit.svg");
  }

  .loading {
    width: 20px;
    height: 20px;
    display: block;
    background-size: 20px 20px;
    background-image: url("../../../assets/images/loading.gif");
    margin-left: 15px;
  }
}
