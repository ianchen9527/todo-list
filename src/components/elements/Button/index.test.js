import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import Button from "./index";

describe("Button", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Button />);
  });

  describe("default props", () => {
    it("should render without crashing", () => {
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe("props.disabled", () => {
    let disabled;
    describe("is true", () => {
      beforeEach(() => {
        disabled = true;
        wrapper.setProps({ disabled });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });

    describe("is false", () => {
      beforeEach(() => {
        disabled = false;
        wrapper.setProps({ disabled });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });
  });

  describe("props.loading", () => {
    let loading;
    describe("is true", () => {
      beforeEach(() => {
        loading = true;
        wrapper.setProps({ loading });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });

    describe("is false", () => {
      beforeEach(() => {
        loading = false;
        wrapper.setProps({ loading });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });
  });

  describe("props.icon", () => {
    let icon;
    describe("is set", () => {
      beforeEach(() => {
        icon = "icon";
        wrapper.setProps({ icon });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });
  });

  describe("props.text", () => {
    let text;
    describe("is set", () => {
      beforeEach(() => {
        text = "text";
        wrapper.setProps({ text });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });
  });

  describe("props.onClick", () => {
    let onClick;
    beforeEach(() => {
      onClick = jest.fn();
      wrapper.setProps({ onClick });
    });

    afterEach(() => {
      onClick.mockClear();
    });

    describe("when .wrapper is clicked", () => {
      beforeEach(() => {
        wrapper.find(".wrapper").simulate("click");
      });

      it("should be triggered", () => {
        expect(onClick).toHaveBeenCalled();
      });
    });
  });
});
