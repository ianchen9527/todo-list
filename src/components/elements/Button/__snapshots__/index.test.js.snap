// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default props should render without crashing 1`] = `
<div
  className="wrapper"
  onClick={null}
/>
`;

exports[`Button props.disabled is false should render without crashing 1`] = `
<div
  className="wrapper"
  onClick={null}
/>
`;

exports[`Button props.disabled is true should render without crashing 1`] = `
<div
  className="wrapper disabled"
  onClick={null}
/>
`;

exports[`Button props.icon is set should render without crashing 1`] = `
<div
  className="wrapper"
  onClick={null}
>
  <div
    className="icon-icon"
  />
</div>
`;

exports[`Button props.loading is false should render without crashing 1`] = `
<div
  className="wrapper"
  onClick={null}
/>
`;

exports[`Button props.loading is true should render without crashing 1`] = `
<div
  className="wrapper"
  onClick={null}
>
  <div
    className="loading"
  />
</div>
`;

exports[`Button props.text is set should render without crashing 1`] = `
<div
  className="wrapper"
  onClick={null}
>
  text
</div>
`;
